storage_controller/src/operation_utils.rs

```diff
@@ -46,11 +46,36 @@ impl TenantShardDrain {
         &self,
         tenants: &BTreeMap<TenantShardId, TenantShard>,
         scheduler: &Scheduler,
-    ) -> Option<NodeId> {
-        let tenant_shard = tenants.get(&self.tenant_shard_id)?;
+    ) -> TenantShardDrainAction {
+        let tenant_shard = match tenants.get(&self.tenant_shard_id) {
+            Some(tenant_shard) => tenant_shard,
+            None => {
+                return TenantShardDrainAction::Skip;
+            }
+        };
 
         if *tenant_shard.intent.get_attached() != Some(self.drained_node) {
-            return None;
+            // If the intent attached node is not the drained node, check the observed state
+            // of the shard on the drained node. If it is Attached*, it means the shard is
+            // beeing migrated from the drained node. The drain loop needs to wait for the
+            // reconciliation to complete for a smooth draining.
+
+            let attach_mode = tenant_shard
+                .observed
+                .locations
+                .get(&self.drained_node)
+                .and_then(|observed| observed.conf.as_ref().map(|conf| conf.mode));
+
+            use pageserver_api::models::LocationConfigMode::{
```

Comment:
if you do an ad hoc use like this, i think it's more common to have an asterisk instead of explicitly listing the variants. plus, it should be at the top of the block.